PublicMapping/districtbuilder
# Allow copying maps

## Overview

If a user is viewing a map (either a shared map or their own), we want to enable them to copy the map as a template and start a new map from it. This will create a copy of the map in the user's account so they can begin editing the map. This also takes into account user login/registration and will ensure a user is logged in before allowing copying.

### Checklist


### Demo

![copy-map-01-fixed](https://user-images.githubusercontent.com/6386/100763681-7f298980-33c3-11eb-9088-7db9c308814a.png)
![copy-map-02](https://user-images.githubusercontent.com/6386/100760806-286e8080-33c0-11eb-9b7a-25e46e86ac68.png)
![copy-map-03](https://user-images.githubusercontent.com/6386/100760807-286e8080-33c0-11eb-875d-618e1089a904.png)
![copy-map-04](https://user-images.githubusercontent.com/6386/100760808-29071700-33c0-11eb-9db8-acd84e68035a.png)

### Notes

 * I based this off the wireframes, which had some placeholder copy. There will need to be a design pass to flesh this out, as well as tweak styling as needed.
 * The acceptance criteria on the issue is a little out-of-date. Read all the comments, and see the wireframes for the full picture. I believe I've addressed everything.
 * Jeff requested that we build the auth modal in such a way that it'd be possible to update the header content/etc. independently from that of the login/registration pages. That is why I took the approach I did here: allowing the ability to pass arbitrary child nodes for rendering, and abstracting out only the bare minimum required fields/functionality from the login/registration pages. This allows keeping things flexible while also eliminating the need for copy/pasted code.
 * There was no mention of adding the password reset functionality to the modal, and in the interest of time I didn't do so. This will be straightforward to implement if we'd like to do so, however.

## Testing Instructions

- `./scripts/server`
- Log in, view a map, and ensure you're able to make a copy of it
- Grab the url for the map, log out, and view the read-only version
- Ensure you're able to make a copy of it (along the way, testing both the login and register functionality of the modal)
- Ensure the login and register functionality of the non-modal pages still work

Closes #493 
Closes #497
